rfcs/migrate-to-async-hyper.md

```diff
@@ -0,0 +1,47 @@
+# Migrating docs.rs to asynchronous code and hyper
+
+## Approach
+
+1. Update `tokio`, `futures` and `rusoto_{s3, core, credential}` to their latest versions
+
+2. Make `main` an `async fn` with `#[tokio::main]` to start the tokio runtime, add the `hyper` dependency and migrate the side services to async (By side services I mean the ones in the daemon)
+
+3. Move all routing to `hyper`, separate the handlers out from being within `iron` code and forward requests with `spawn_blocking`
+   1. Routes should be migrated to asynchronous code from least used to most used
```

Comment:
Rather than moving all routing to `hyper` at once, the `hyper` router could just have a catch-all final route that forwards into the `iron` router.

Reply:
I also considered that, but I think that moving routing would be the easiest option, as the way we have things set up makes it so that every route is a potential crate, so forwarding would be very difficult to implement well. As mentioned, we already have custom routing code in place, so moving over shouldn't be to large of a hassle

Reply:
> I also considered that, but I think that moving routing would be the easiest option, as the way we have things set up makes it so that every route is a potential crate, so forwarding would be very difficult to implement well.

I don't see why this is the case? If we recognize the route, send it to hyper, otherwise send it to iron. If hyper recognizes the route as a crate it will serve the request, otherwise it won't recognize it and will forward to iron. We just have to make sure we don't add new top-level routes in hyper.

Reply:
The other reason that that's not an ideal solution is that it breaks the flow of async, so we'll have to do a lot more work to get hyper's responses back onto the runtime. Having hyper determine things also makes it easier to migrate routes, as it only affects the route code, so multiple routes can be migrated in separate PRs without conflicting. Using hyper to forward requests is just much easier of a solution from my point of view

Reply:
`iron` and `hyper` are both `fn(Request) -> Response` based aren't they? So you'd just have a final handler like

```rust
get("*", |request| async move {
  tokio::blocking(|| {
    let request = map_to_iron(request);
    let response = iron_router.handle(request);
    map_to_hyper(response)
  }.await
});
```

Reply:
Oh, I think I understand what y'all are proposing, I guess that's a viable solution

Reply:
Please update the RFC to reflect this alternative.